Extension/src/LanguageServer/client.ts

```diff
@@ -985,7 +1074,18 @@ export class DefaultClient implements Client {
                         // Listen for messages from the language server.
                         this.registerNotifications();
                     } else {
+                        // update all client configurations
                         this.configuration.CompilerDefaults = compilerDefaults;
+                        this.configuration.setupConfigurations();
```

Comment:
This code seems duplicated.  That can be an issue, if a bug is fixed in one place and not the other.  Could you reorganize this code so the duplication is not required?  In this case, it looks like some code that applies to both conditions can be moved to after the `else` block.

Reply:
I don't understand, it looks like it should have to go before the isFirstClient check correct?

Reply:
There are actually a couple problems here.  Because all of this is within a lambba passed to `this.queueBlockingTask`, the logic within it is not synchronized.  By that I mean: As soon as you call an async function (i.e. awaiting this.requestCompiler), it's possible for the lambda's of other clients to start executing.  This means that `compilerDefaults`, which you are referencing in the else block, may not yet actually be populated.  You'll need to add some synchronization here, such as setting aside the Promise returned by requestCompiler and also awaiting it in the else block and getting the compilerDefaults out of that promise.

Back to the code duplication issue: Since you have some logic (the call to requestCompiler) that needs to happen first, it would seem to make more sense to move the common/duplicate code to after the else block.  (There is no reason the call to promptSelectCompiler can't happen after the call to this.registerNotifications.